### PR TITLE
(DOCSP-26946): React Native Unit Test Suite

### PR DESCRIPTION
**Pull request by mongomoe**
### Jira

- https://jira.mongodb.org/browse/DOCSP-26946

### Staged Changes

N/A

### Reminder Checklist

If your PR modifies the docs, you might need to also update some corresponding
pages. Check if completed or N/A.

- [N/A] Create Jira ticket for corresponding docs-app-services update(s), if any
- [N/A] Checked/updated Admin API
- [N/A] Checked/updated CLI reference

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: examples/reactnative/__tests__/CRUD/create-test.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import React from "react";
+import { Button, TextInput } from "react-native";
+import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
+import Realm from "realm";
+import { createRealmContext } from '@realm/react'
+
+class Dog extends Realm.Object {
```
Comment (mongomoe) [inline examples/reactnative/__tests__/CRUD/create-test.tsx:7]: FYI to the reviewer: the TS schemas currently use 'static schema' but will be replaced to use the new Realm Babel Plugin which has a more native/intuitive TS like syntax and doesn't require a static schema. Just waiting on a release of the plugin, that will fix a sync bug + allow us to utilize the plugin.

File: examples/reactnative/__tests__/CRUD/read-test.tsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
+import React from "react";
+import { View, Text } from "react-native";
+import { render, waitFor} from "@testing-library/react-native";
+import Realm from "realm";
+import { createRealmContext } from '@realm/react'
+
+class Person extends Realm.Object {
```
Comment (mongomoe) [inline examples/reactnative/__tests__/CRUD/read-test.tsx:7]: FYI to the reviewer: the TS schemas currently use 'static schema' but will be replaced to use the new Realm Babel Plugin which has a more native/intuitive TS like syntax and doesn't require a static schema. Just waiting on a release of the plugin, that will fix a sync bug + allow us to utilize the plugin.

File: examples/reactnative/README.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
+# Realm Examples
```
Comment (cbush) [inline examples/reactnative/README.md:1]: To be consistent with the source/sdk/react-native, it would be nice to rename the directory to `examples/react-native`

File: examples/reactnative/README.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+| `testSetup.js`     | Setup and cleanup for the Jest tests. Runs immediately after Jest loads.                |
+
+
+## Develop
+
+### Create a New Test Case File
+
+The file structure in `__tests__/` is intended to match the IA structure in the
+React Native docs. If you're adding a new code example to an existing page, add
+an additional it/test block within the existing page.
```
Comment (cbush) [inline examples/reactnative/README.md:57]:
```suggestion
an additional it/test block within the test file corresponding to that page.
```

File: examples/reactnative/README.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+If you're adding a new section in the TOC of the docs, create a
+new folder under `__tests__`. For each page in that section, create a test file
+within the new folder.
+
+However, if you're just adding a new page, create a new page within the relevant
+subfolder of `__test__`. For example, if you're creating examples for a new page
+called "read.txt" that is under the "CRUD" section of the TOC in the docs, you
+should create the file `__tests__/CRUD/read.tsx|jsx`. That new file should have
+it's own describe block labelled accordingly to the page.
```
Comment (cbush) [inline examples/reactnative/README.md:67]:
```suggestion
its own describe block labelled accordingly to the page.
```

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import React from "react";
+import { Button, TextInput } from "react-native";
```
Comment (cbush) [inline examples/reactnative/__tests__/js/CRUD/create-test.jsx:2]: Noticing that the formatting doesn't match the prettier configuration. The problem is that anyone editing this file after you will have a bunch of unrelated changes in their commits when their formatter fixes everything. You should have your VS Code configured to run prettier format on save using the .prettierrc in the project directory.

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        const wrapper = ({ children }) => <RealmProvider>{children}</RealmProvider>;
+
+        // :snippet-start: crud-create-object
+        // :replace-start: {
+        //  "terms": {
+        //   " testID='handleAddItemBtn'": ""
+        //   }
+        // }
+        const CreateDogInput = () => {
+            const [dogName, onChangeDogName] = React.useState('Fido');
```
Comment (cbush) [inline examples/reactnative/__tests__/js/CRUD/create-test.jsx:60]: Typically you would call these [x, setX]
```suggestion
            const [dogName, setDogName] = React.useState('Fido');
```

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        assertionRealm = await Realm.open(realmConfig)
+
+        // delete every object in the realmConfig in the Realm to make test idempotent
+        assertionRealm.write(()=> {
+            assertionRealm.delete(assertionRealm.objects("Dog"));
+            assertionRealm.delete(assertionRealm.objects("Person"));
+        })
+    })
+    it("should create a new object", async () => {        
+        const wrapper = ({ children }) => <RealmProvider>{children}</RealmProvider>;
```
Comment (cbush) [inline examples/reactnative/__tests__/js/CRUD/create-test.jsx:51]: Is this used?

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            const handleAddItem = () => {
+                realm.write(() => {
+                    new Dog(realm, {  name: dogName, age: 1, })
+                });
+            }
+
+            return (
+                <>
+                    <TextInput onChangeText={onChangeDogName} value={dogName} />
+                    <Button onPress={() => handleAddItem()} title="Submit Item" testID='handleAddItemBtn'/>
```
Comment (cbush) [inline examples/reactnative/__tests__/js/CRUD/create-test.jsx:72]: It seems like we're mixing metaphors here. We have "add", "create", and "submit" both "item" and "dog". It would be a lot clearer if this were simplified to one consistent thing.

File: examples/reactnative/README.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+**TypeScript**
+
+```bash
+npm run test
+```
+
+**JavaScript**
+
+todo: Add JavaScript support
```
Comment (cbush) [inline examples/reactnative/README.md:36]: Will this be done as part of this PR or later?

**Review by cbush — COMMENTED** [review]
I understand this is a WIP. Not sure which parts are really up for review or not. So I can't approve. Generally the setup looks ok though!

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx (hunk 6 above)
Comment (mongomoe) [reply examples/reactnative/__tests__/js/CRUD/create-test.jsx]: Already fixed in https://github.com/mongodb/docs-realm/pull/2381 
For this PR, you can mostly just ignore any specific issues with the code in __tests__, just want to make sure it runs + is being tested properly.

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx (hunk 7 above)
Comment (mongomoe) [reply examples/reactnative/__tests__/js/CRUD/create-test.jsx]: Fixed in the latest commit of https://github.com/mongodb/docs-realm/pull/2381

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx (hunk 8 above)
Comment (mongomoe) [reply examples/reactnative/__tests__/js/CRUD/create-test.jsx]: Nope, it's already removed in https://github.com/mongodb/docs-realm/pull/2381, this file (as well as the other __tests__/**/*.jsx|**/*.tsx files) will be removed from this PR before I merge this either way so I'll keep it as is for now.

File: examples/reactnative/__tests__/js/CRUD/create-test.jsx (hunk 9 above)
Comment (mongomoe) [reply examples/reactnative/__tests__/js/CRUD/create-test.jsx]: Good call, just fixed now in a [commit](https://github.com/mongodb/docs-realm/pull/2381/commits/0156d3338cc0e2853f49f4ae0c8349642cb1a2c4) of https://github.com/mongodb/docs-realm/pull/2381

File: examples/reactnative/README.md (hunk 10 above)
Comment (mongomoe) [reply examples/reactnative/README.md]: I've just removed this line, and changed this whole paragraph to:

### Run Tests

This project defines a unit test suite that tests both JS and TS code examples, run them via:

```bash
npm run test
```

Rather than run the tests separately with `npm run test:js` or `npm run test:ts` like we do in the Node.js unit test suite.

**Comment by mongomoe** [comment]
@cbush were you able to run the tests locally and get them passing in your local environment?

**Comment by cbush** [comment]
On first-ever run there was a fail in the create test where the button hadn't rendered before the test tried to get it by test id... but I couldn't repro after that.

File: .github/workflows/react-native.yml [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  build:
+    name: Run React Native Example App Tests
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x]
+
+    steps:
+    - uses: actions/checkout@v3
```
Comment (cbush) [inline .github/workflows/react-native.yml:18]: Formatting: for consistency, from line 18-24 you need two extra spaces of indentation. Compare with line 5-6.

 Btw, there is a plugin for VS Code that will auto-format for  you. Really handy!

**Review by cbush — APPROVED** [review]
Thanks!
